File: api/v1/schemas/intraday.py
# -*- coding: utf-8 -*-
"""分时做T API 响应模型"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class IntradayKlinePoint(BaseModel):
    """分时K线数据点"""

    Open: float
    High: float
    Low: float
    Close: float
    Volume: float
    Amount: Optional[float] = None
    AvgPrice: Optional[float] = None  # 累计分时均价（成交额/成交量）
    timestamp: str  # ISO格式时间字符串


class WeightContribution(BaseModel):
    """单项权重贡献明细"""

    key: str  # 权重键名
    label: str  # 中文标签
    weight: float  # 满分数
    triggered: bool = False  # 是否触发
    score: float = 0.0  # 实际得分


class IntradaySignal(BaseModel):
    """做T信号"""

    stock_code: str
    signal_type: str  # "buy" | "sell"
    trigger_time: str
    price: float
    score: float
    max_score: int = 10
    confidence: float
    position_advice: str
    reasoning: str = ""
    gravity_adjustment: float = 0.0  # 引力场修正分
    support_force: float = 0.0
    pressure_force: float = 0.0
    buy_weight_details: List[WeightContribution] = Field(default_factory=list)
    sell_weight_details: List[WeightContribution] = Field(default_factory=list)


class ReferenceLine(BaseModel):
    """支撑/压力参考线"""

    id: str
    label: str
    price: float
    category: str
    color: str
    style: str = "dashed"
    base_weight: float = 1.0


# ── 指标子图模型 ──


class IndicatorLinePoint(BaseModel):
    """指标线数据点"""

    time: str  # 时间标签 "HH:MM"
    value: float


class IndicatorLine(BaseModel):
    """指标图中的一条线"""

    name: str
    label: str
    color: str
    data: List[IndicatorLinePoint] = Field(default_factory=list)


class IndicatorSubChart(BaseModel):
    """单个指标子图"""

    id: str  # "absorption" / "main_in_out" / "cyw"
    label: str  # 中文标题 "主力吸筹"
    height: int = 120
    lines: List[IndicatorLine] = Field(default_factory=list)
    signal_text: str = ""  # 最新信号文本，如 "正T买入 ↑"、"卖出 ↓" 等
    metadata: Optional[Dict[str, Any]] = Field(
        default=None,
        description="指标元数据，如MACD的bar_sum/bar_diff，与后端策略算法一致",
    )


class IntradayDataResponse(BaseModel):
    """分时数据完整响应"""

    stock_code: str
    stock_name: str = ""
    date: str
    latest_price: float = Field(default=0.0, description="最新价（用于搜索历史快照回填）")
    change_pct: float = Field(default=0.0, description="涨跌幅（用于搜索历史快照回填）")
    kline_data: List[IntradayKlinePoint] = Field(default_factory=list)
    signals: List[IntradaySignal] = Field(default_factory=list)
    reference_lines: List[ReferenceLine] = Field(default_factory=list)
    indicator_sub_charts: List[IndicatorSubChart] = Field(default_factory=list)
    signal_summary: Dict[str, Any] = Field(default_factory=dict)
    rsi_overbought: float = Field(default=65, description="RSI超买阈值（来自策略配置）")
    rsi_oversold: float = Field(default=20, description="RSI超卖阈值（来自策略配置）")
    mfi_overbought: float = Field(default=80, description="MFI超买阈值（来自策略配置）")
    mfi_oversold: float = Field(default=20, description="MFI超卖阈值（来自策略配置）")
    buy_weights: Dict[str, float] = Field(default_factory=dict, description="买入信号权重配置（来自策略配置）")
    sell_weights: Dict[str, float] = Field(default_factory=dict, description="卖出信号权重配置（来自策略配置）")
    warm_up_summary: Optional[Dict[str, Any]] = Field(
        default=None,
        description="前一个交易日的分时终值快照，用于开盘时指标预热（含价格、成交量、技术指标终值）",
    )
    warmup_info: Optional[Dict[str, Any]] = Field(
        default=None,
        description="预热状态信息: {last_klines_count, prev_date, enabled}",
    )


class SearchHistoryItem(BaseModel):
    """搜索历史条目"""

    id: int
    stock_code: str
    stock_name: str = ""
    date: str = ""
    search_time: str = ""


class SearchHistoryResponse(BaseModel):
    """搜索历史响应"""

    items: List[SearchHistoryItem] = Field(default_factory=list)
    total: int = 0


class SearchHistoryRequest(BaseModel):
    """保存搜索历史请求"""

    stock_code: str
    stock_name: str = ""
    date: str = ""


class DeleteHistoryResponse(BaseModel):
    """删除历史响应"""

    success: bool = True
    message: str = ""


# ── 批量状态查询 ──


class StockSnapshot(BaseModel):
    """单只股票实时快照（腾讯 qt.gtimg.cn 数据源，含五档盘口）"""

    stock_code: str
    stock_name: str = ""
    latest_price: float = 0.0
    change_pct: float = 0.0  # 涨跌幅百分比
    open_price: float = 0.0
    high: float = 0.0
    low: float = 0.0
    pre_close: float = 0.0  # 昨收价
    volume: int = 0  # 成交量(股)
    timestamp: str = ""  # "HH:MM:SS"

    # 五档盘口（卖1→卖5价/量，买1→买5价/量，量单位为手）
    ask_prices: List[float] = Field(default_factory=list)  # [卖一价, 卖二价, 卖三价, 卖四价, 卖五价]
    ask_volumes: List[int] = Field(default_factory=list)  # [卖一量, 卖二量, 卖三量, 卖四量, 卖五量] (手)
    bid_prices: List[float] = Field(default_factory=list)  # [买一价, 买二价, 买三价, 买四价, 买五价]
    bid_volumes: List[int] = Field(default_factory=list)  # [买一量, 买二量, 买三量, 买四量, 买五量] (手)

    # 估值指标（可选，腾讯接口附带）
    volume_ratio: Optional[float] = None  # 量比
    turnover_rate: Optional[float] = None  # 换手率(%)
    pe_ratio: Optional[float] = None  # 市盈率
    pb_ratio: Optional[float] = None  # 市净率


class BatchStatusRequest(BaseModel):
    """批量状态查询请求"""

    stock_codes: List[str] = Field(..., description="需要查询的股票代码列表")
    current_code: str = Field("", description="当前展示的股票代码")
    include_signals: bool = Field(False, description="是否对全部股票进行信号检测（用于铃铛通知）")
    skip_kline_fetch: bool = Field(False, description="是否跳过K线拉取（仅返回快照，可视化页面使用）")
    existing_kline_count: int = Field(0, description="前端已有的K线数量，用于增量返回新K线（0表示返回全部）")


class SignalAlert(BaseModel):
    """单只股票的信号告警"""

    stock_code: str
    signal_type: str  # "buy" | "sell"
    trigger_time: str  # 信号触发时间
    price: float  # 信号触发价


class BatchStatusResponse(BaseModel):
    """批量状态查询响应"""

    snapshots: Dict[str, StockSnapshot] = Field(default_factory=dict)
    current_updated: bool = False  # 当前展示股票是否有新数据
    current_full_data: Optional[IntradayDataResponse] = None  # 仅 current_updated=True 时有值
    signal_alerts: Optional[Dict[str, Optional[SignalAlert]]] = Field(
        default=None,
        description="当 include_signals=True 时返回每只股票的最新信号告警",
    )


class SimulatedTradeItem(BaseModel):
    """单笔模拟交易记录（摘要）"""

    buy_time: str = ""
    buy_price: float = 0.0
    sell_time: str = ""
    sell_price: float = 0.0
    return_pct: float = 0.0


class SimulationReportResponse(BaseModel):
    """模拟交易统计报告"""

    stock_code: str = ""
    total_klines: int = 0
    total_signals: int = 0
    buy_signals: int = 0
    sell_signals: int = 0
    total_trades: int = 0
    win_trades: int = 0
    lose_trades: int = 0
    win_rate: float = 0.0
    avg_return_pct: float = 0.0
    max_return_pct: float = 0.0
    min_return_pct: float = 0.0
    total_return_pct: float = 0.0
    max_drawdown_pct: float = 0.0
    profit_factor: float = 0.0
    trades: List[SimulatedTradeItem] = Field(default_factory=list)


class BatchDownloadStatus(BaseModel):
    """批量下载分时数据进度"""

    task_id: str = ""
    status: str = "idle"  # idle | running | completed | cancelled | failed
    total: int = 0
    completed: int = 0
    failed: int = 0
    skipped: int = 0
    current_code: str = ""
    current_name: str = ""
    elapsed_seconds: float = 0.0
    errors: List[dict] = Field(default_factory=list)  # [{"code": ..., "error": ...}], 最多 20 条
    date: str = ""
    paused: bool = False
    waiting_retry: bool = False
    retry_countdown: int = 0


class FailedListItem(BaseModel):
    """批量下载失败标的项"""

    code: str = ""
    error_msg: str = ""
    retry_count: int = 0


class FailedListResponse(BaseModel):
    """批量下载失败列表响应"""

    date: str = ""
    failed_list: List[FailedListItem] = Field(default_factory=list)
    count: int = 0

File: apps/dsa-web/src/api/intraday.ts
/**
 * 分时做T API 模块
 *
 * 封装分时K线数据获取、做T信号和参考线的请求接口。
 */

const API_BASE = '/api/v1/intraday';

// ============================================================
// TypeScript 类型定义
// ============================================================

export interface IntradayKlinePoint {
  Open: number;
  High: number;
  Low: number;
  Close: number;
  Volume: number;
  Amount?: number;
  AvgPrice?: number;
  timestamp: string;
  time?: string;
}

export interface WeightContribution {
  key: string;
  label: string;
  weight: number;
  triggered: boolean;
  score: number;
}

export interface IntradaySignal {
  stock_code: string;
  signal_type: 'buy' | 'sell' | 'hold';
  trigger_time: string;
  price: number;
  score: number;
  max_score: number;
  confidence: number;
  position_advice: string;
  reasoning: string;
  gravity_adjustment?: number;
  support_force?: number;
  pressure_force?: number;
  buy_weight_details?: WeightContribution[];
  sell_weight_details?: WeightContribution[];
}

export interface ReferenceLine {
  id: string;
  label: string;
  price: number;
  category: string;
  color: string;
  style: string;
  base_weight: number;
}

export interface IndicatorLinePoint {
  time: string;
  value: number;
}

export interface IndicatorLine {
  name: string;
  label: string;
  color: string;
  data: IndicatorLinePoint[];
}

export interface IndicatorSubChart {
  id: string;
  label: string;
  height: number;
  lines: IndicatorLine[];
  signal_text: string;
  metadata?: Record<string, any> | null;
}

export interface FiveMinKlinePoint {
  Open: number;
  High: number;
  Low: number;
  Close: number;
  Volume: number;
  timestamp: string;
}

export interface TiandaoSignal {
  signal_type: 'buy' | 'sell';
  trigger_time: string;
  price: number;
  reason: string;
}

export interface TiandaoSubChart {
  klines: FiveMinKlinePoint[];
  prev_day_klines: FiveMinKlinePoint[];
  jinzuan_line: IndicatorLinePoint[];
  jinniu_line: IndicatorLinePoint[];
  signals: TiandaoSignal[];
}

export interface IntradayDataResponse {
  stock_code: string;
  stock_name: string;
  date: string;
  kline_data: IntradayKlinePoint[];
  signals: IntradaySignal[];
  reference_lines: ReferenceLine[];
  indicator_sub_charts: IndicatorSubChart[];
  tiandao_sub_chart?: TiandaoSubChart | null;
  signal_summary: {
    buy_signals: number;
    sell_signals: number;
    total_signals: number;
    strong_signals: number;
    medium_signals: number;
    weak_signals: number;
    simulated_return_pct: number;
  };
  rsi_overbought?: number;
  rsi_oversold?: number;
  mfi_overbought?: number;
  mfi_oversold?: number;
  buy_weights?: Record<string, number>;
  sell_weights?: Record<string, number>;
  warm_up_summary?: Record<string, any> | null;
  warmup_info?: {
    enabled: boolean;
    last_klines_count: number;
    prev_date: string;
    klines?: IntradayKlinePoint[];
  } | null;
}

export interface SearchHistoryItem {
  id: number;
  stock_code: string;
  stock_name: string;
  date: string;
  search_time: string;
}

export interface SearchHistoryResponse {
  items: SearchHistoryItem[];
  total: number;
}

export interface StockSnapshot {
  stock_code: string;
  stock_name: string;
  latest_price: number;
  change_pct: number;
  open_price: number;
  high: number;
  low: number;
  pre_close: number;
  volume: number;
  timestamp: string;
  // 五档盘口（卖1→卖5，买1→买5，量单位为手）
  ask_prices: number[];    // [卖一价, 卖二价, 卖三价, 卖四价, 卖五价]
  ask_volumes: number[];   // [卖一量, 卖二量, 卖三量, 卖四量, 卖五量] (手)
  bid_prices: number[];    // [买一价, 买二价, 买三价, 买四价, 买五价]
  bid_volumes: number[];   // [买一量, 买二量, 买三量, 买四量, 买五量] (手)
  // 估值指标（可选）
  volume_ratio?: number | null;
  turnover_rate?: number | null;
  pe_ratio?: number | null;
  pb_ratio?: number | null;
}

export interface SignalAlert {
  stock_code: string;
  signal_type: 'buy' | 'sell';
  trigger_time: string;
  price: number;
}

export interface BatchStatusResponse {
  snapshots: Record<string, StockSnapshot>;
  current_updated: boolean;
  current_full_data: IntradayDataResponse | null;
  signal_alerts?: Record<string, SignalAlert | null> | null;
}

export interface BatchStatusRequest {
  stock_codes: string[];
  current_code: string;
  include_signals?: boolean;
}

export interface SimulatedTradeItem {
  buy_time: string;
  buy_price: number;
  sell_time: string;
  sell_price: number;
  return_pct: number;
}

export interface SimulationReportResponse {
  stock_code: string;
  total_klines: number;
  total_signals: number;
  buy_signals: number;
  sell_signals: number;
  total_trades: number;
  win_trades: number;
  lose_trades: number;
  win_rate: number;
  avg_return_pct: number;
  max_return_pct: number;
  min_return_pct: number;
  total_return_pct: number;
  max_drawdown_pct: number;
  profit_factor: number;
  trades: SimulatedTradeItem[];
}

export interface TradingStatus {
  is_trading_day: boolean;
  is_trading_time: boolean;
  next_session_start: string | null;
}

export interface IntradayConfig {
  polling_interval_ms: number;
  batch_download_polling_interval_ms: number;
  screen_async_polling_interval_ms: number;
  signal_sound_cooldown_seconds: number;
  signal_sound_volume: number;
  trading_status: TradingStatus;
}

// ============================================================
// API 函数
// ============================================================

export async function getIntradayConfig(): Promise<IntradayConfig> {
  const defaultConfig: IntradayConfig = {
    polling_interval_ms: 30000,
    batch_download_polling_interval_ms: 1000,
    screen_async_polling_interval_ms: 1000,
    signal_sound_cooldown_seconds: 30,
    signal_sound_volume: 0.3,
    trading_status: {
      is_trading_day: false,
      is_trading_time: false,
      next_session_start: null,
    },
  };
  const resp = await fetch(`${API_BASE}/config`);
  if (!resp.ok) return defaultConfig;
  return resp.json();
}

export async function getIntradayData(
  stockCode: string,
  date?: string,
  strategy?: string,
  warmupEnabled?: boolean,
): Promise<IntradayDataResponse> {
  const params = new URLSearchParams();
  if (date) {
    params.set('date', date.replace(/-/g, ''));
  }
  if (strategy) {
    params.set('strategy', strategy);
  }
  if (warmupEnabled !== undefined) {
    params.set('warmup_enabled', String(warmupEnabled));
  }
  const query = params.toString();
  const url = `${API_BASE}/data/${stockCode}${query ? `?${query}` : ''}`;

  const resp = await fetch(url);
  if (!resp.ok) {
    const error = await resp.json().catch(() => ({ message: resp.statusText }));
    throw new Error(error.detail?.message || error.message || `请求失败 (${resp.status})`);
  }
  return resp.json();
}

export async function getSearchHistory(
  limit: number = 20,
): Promise<SearchHistoryResponse> {
  const resp = await fetch(`${API_BASE}/history?limit=${limit}`);
  if (!resp.ok) return { items: [], total: 0 };
  return resp.json();
}

export async function saveSearchHistory(
  stockCode: string,
  stockName: string = '',
  date: string = '',
): Promise<SearchHistoryItem> {
  const resp = await fetch(`${API_BASE}/history`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ stock_code: stockCode, stock_name: stockName, date }),
  });
  if (!resp.ok) throw new Error('保存搜索历史失败');
  return resp.json();
}

export async function deleteSearchHistory(id: number): Promise<boolean> {
  const resp = await fetch(`${API_BASE}/history/${id}`, { method: 'DELETE' });
  return resp.ok;
}

export async function updateSearchHistoryTimestamp(id: number): Promise<SearchHistoryItem> {
  const resp = await fetch(`${API_BASE}/history/${id}/timestamp`, { method: 'PUT' });
  if (!resp.ok) throw new Error('更新时间戳失败');
  return resp.json();
}

export async function getBatchStatus(
  stockCodes: string[],
  currentCode: string,
  includeSignals: boolean = false,
  existingKlineCount: number = 0,
  skipKlineFetch: boolean = false,
): Promise<BatchStatusResponse> {
  const resp = await fetch(`${API_BASE}/batch-status`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      stock_codes: stockCodes,
      current_code: currentCode,
      include_signals: includeSignals,
      existing_kline_count: existingKlineCount,
      skip_kline_fetch: skipKlineFetch,
    }),
  });
  if (!resp.ok) throw new Error(`批量状态查询失败: ${resp.status}`);
  return resp.json();
}

export async function simulateTrading(
  stockCode: string,
): Promise<SimulationReportResponse> {
  const resp = await fetch(`${API_BASE}/${stockCode}/simulate-trading`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  if (!resp.ok) throw new Error(`模拟交易失败: ${resp.status}`);
  return resp.json();
}

// ============================================================
// 批量下载分时数据
// ============================================================

export interface BatchDownloadStatus {
  task_id: string;
  status: 'idle' | 'running' | 'completed' | 'cancelled' | 'failed';
  total: number;
  completed: number;
  failed: number;
  skipped: number;
  current_code: string;
  current_name: string;
  elapsed_seconds: number;
  errors: { code: string; error: string }[];
  date: string;
  paused: boolean;
  waiting_retry: boolean;
  retry_countdown: number;
}

export async function startBatchDownload(
  date?: string,
  maxWorkers?: number,
  force?: boolean,
): Promise<BatchDownloadStatus> {
  const body: Record<string, any> = {};
  if (date) body.date = date;
  if (maxWorkers) body.max_workers = maxWorkers;
  if (force !== undefined) body.force = force;
  const resp = await fetch(`${API_BASE}/batch-download`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!resp.ok) throw new Error(`批量下载启动失败: ${resp.status}`);
  return resp.json();
}

export async function getBatchDownloadStatus(
  taskId?: string,
): Promise<BatchDownloadStatus> {
  // 添加时间戳防止浏览器缓存 GET 请求
  const ts = Date.now();
  const params = taskId
    ? `?task_id=${encodeURIComponent(taskId)}&_=${ts}`
    : `?_=${ts}`;
  const resp = await fetch(`${API_BASE}/batch-download/status${params}`, {
    headers: { 'Cache-Control': 'no-cache', 'Pragma': 'no-cache' },
  });
  if (!resp.ok) throw new Error(`查询进度失败: ${resp.status}`);
  return resp.json();
}

export async function cancelBatchDownload(
  taskId: string,
): Promise<{ message: string }> {
  const resp = await fetch(`${API_BASE}/batch-download/cancel`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ task_id: taskId }),
  });
  if (!resp.ok) throw new Error(`取消失败: ${resp.status}`);
  return resp.json();
}

export async function togglePauseBatchDownload(
  taskId: string,
): Promise<{ message: string; paused: boolean }> {
  const resp = await fetch(`${API_BASE}/batch-download/pause`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ task_id: taskId }),
  });
  if (!resp.ok) throw new Error(`暂停/继续失败: ${resp.status}`);
  return resp.json();
}

export async function retryFailedBatchDownload(
  date?: string,
  maxWorkers?: number,
): Promise<BatchDownloadStatus> {
  const body: Record<string, any> = {};
  if (date) body.date = date;
  if (maxWorkers) body.max_workers = maxWorkers;
  const resp = await fetch(`${API_BASE}/batch-download/retry-failed`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!resp.ok) throw new Error(`失败重试启动失败: ${resp.status}`);
  return resp.json();
}

export interface FailedListItem {
  code: string;
  error_msg: string;
  retry_count: number;
}

export interface FailedListResponse {
  date: string;
  failed_list: FailedListItem[];
  count: number;
}

export async function getBatchDownloadFailedList(
  date?: string,
): Promise<FailedListResponse> {
  const ts = Date.now();
  const params = date
    ? `?date=${encodeURIComponent(date)}&_=${ts}`
    : `?_=${ts}`;
  const resp = await fetch(`${API_BASE}/batch-download/failed-list${params}`, {
    headers: { 'Cache-Control': 'no-cache', 'Pragma': 'no-cache' },
  });
  if (!resp.ok) throw new Error(`查询失败列表失败: ${resp.status}`);
  return resp.json();
}
